File: src/test/java/com/polygloat/Assertions/ErrorResponseAssert.java
package com.polygloat.Assertions;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.core.type.TypeReference;
import com.fasterxml.jackson.databind.ObjectMapper;
import lombok.SneakyThrows;
import org.assertj.core.api.AbstractAssert;
import org.springframework.test.web.servlet.MvcResult;

import java.io.UnsupportedEncodingException;
import java.util.List;
import java.util.Map;

public class ErrorResponseAssert extends AbstractAssert<ErrorResponseAssert, MvcResult> {

    public ErrorResponseAssert(MvcResult mvcResult) {
        super(mvcResult, ErrorResponseAssert.class);
    }

    public StandardValidationMessageAssert isStandardValidation() {
        Map<String, String> standardValidation = getStandardMap().get("STANDARD_VALIDATION");
        if (standardValidation == null) {
            failWithMessage("Error response is not standard validation type.");
        }
        return new StandardValidationMessageAssert(standardValidation);
    }

    public CustomValidationMessageAssert isCustomValidation() {
        Map<String, List<Object>> customValidation = getCustomMap().get("CUSTOM_VALIDATION");
        if (customValidation == null) {
            failWithMessage("Error response is not custom validation type.");
        }
        return new CustomValidationMessageAssert(customValidation);
    }

    private Map<String, Map<String, String>> getStandardMap() {
        ObjectMapper objectMapper = new ObjectMapper();
        try {
            return objectMapper.readValue(actual.getResponse().getContentAsString(), new TypeReference<>() {
            });
        } catch (JsonProcessingException | UnsupportedEncodingException e) {
            throw new RuntimeException("Can not parse error response.");
        }
    }

    @SneakyThrows
    private Map<String, Map<String, List<Object>>> getCustomMap() {
        ObjectMapper objectMapper = new ObjectMapper();
        try {
            return objectMapper.readValue(actual.getResponse().getContentAsString(), new TypeReference<>() {
            });
        } catch (JsonProcessingException | UnsupportedEncodingException e) {
            throw new RuntimeException("Can not parse error response:\n" + actual.getResponse().getContentAsString());
        }
    }
}

File: backend/testing/src/main/kotlin/io/tolgee/testing/assertions/ErrorResponseAssert.kt
package io.tolgee.testing.assertions

import com.fasterxml.jackson.core.JsonProcessingException
import io.tolgee.fixtures.mapResponseTo
import io.tolgee.testing.assertions.Assertions.assertThat
import org.assertj.core.api.AbstractAssert
import org.springframework.test.web.servlet.MvcResult
import java.io.UnsupportedEncodingException

class ErrorResponseAssert(
  mvcResult: MvcResult?,
) : AbstractAssert<ErrorResponseAssert?, MvcResult?>(mvcResult, ErrorResponseAssert::class.java) {
  val isStandardValidation: StandardValidationMessageAssert
    get() {
      val standardValidation = standardMap["STANDARD_VALIDATION"]
      if (standardValidation == null) {
        failWithMessage("Error response is not standard validation type.")
      }
      return standardValidation?.let { StandardValidationMessageAssert(it) }!!
    }
  val isCustomValidation: CustomValidationMessageAssert
    get() {
      val customValidation = customMap["CUSTOM_VALIDATION"]
      if (customValidation == null) {
        failWithMessage("Error response is not custom validation type.")
      }
      return CustomValidationMessageAssert(customValidation)
    }

  fun hasCode(code: String) {
    val content: Map<String, Any> = actual!!.mapResponseTo()
    assertThat(content["code"]).isEqualTo(code)
  }

  private val standardMap: Map<String, Map<String, String>>
    get() {
      return try {
        actual!!.mapResponseTo()
      } catch (e: JsonProcessingException) {
        throw RuntimeException("Can not parse error response.")
      } catch (e: UnsupportedEncodingException) {
        throw RuntimeException("Can not parse error response.")
      }
    }
  private val customMap: Map<String, Map<String, List<Any>>>
    get() {
      return try {
        actual!!.mapResponseTo()
      } catch (e: JsonProcessingException) {
        try {
          throw RuntimeException(
            """
            Can not parse error response:
            ${actual!!.response.contentAsString}
            """.trimIndent(),
          )
        } catch (unsupportedEncodingException: UnsupportedEncodingException) {
          throw RuntimeException(unsupportedEncodingException)
        }
      } catch (e: UnsupportedEncodingException) {
        try {
          throw RuntimeException(
            """
            Can not parse error response:
            ${actual!!.response.contentAsString}
            """.trimIndent(),
          )
        } catch (unsupportedEncodingException: UnsupportedEncodingException) {
          throw RuntimeException(unsupportedEncodingException)
        }
      }
    }
}
